### PR TITLE
MAVSDK tests: Use full system timeout off

### PR DESCRIPTION
**Pull request by LorenzMeier**
Depending on loop iterations for timeouts is not accurate, as usleep behavior depends on the system load and asking for a 0 ms sleep can potentially return immediately.

### CODE REVIEW

File: test/mavsdk_tests/autopilot_tester.h [hunk 1]
```diff
@@ -128,24 +128,27 @@ class AutopilotTester
 			uint32_t start_time = _info->get_flight_information().second.time_boot_ms;
 
 			while (!fun()) {
-				std::this_thread::sleep_for(duration_ms / 1000);
+				std::this_thread::sleep_for(duration_ms / 100);
```
Comment (julianoes) [inline test/mavsdk_tests/autopilot_tester.h:132]: This is wrong now.

**Review by julianoes — CHANGES_REQUESTED** [review]
More thorough review coming later.

File: test/mavsdk_tests/autopilot_tester.h (hunk 1 above)
Comment (LorenzMeier) [reply test/mavsdk_tests/autopilot_tester.h:132]: It is not - the sleep quantum is completely independent (the way it is implemented now).

**Comment by LorenzMeier** [comment]
I suspect that while I consider the solution with a global timeout better than using sleep to time, it is not material. I'm collecting observations here:
https://github.com/PX4/PX4-Autopilot/issues/16464

File: test/mavsdk_tests/autopilot_tester.h [hunk 2]
```diff
@@ -128,24 +128,31 @@ class AutopilotTester
...
 		} else {
 			// Nothing is connected yet. Use the host time.
-			unsigned iteration = 0;
+			const auto start_time = std::chrono::steady_clock::now();
 
 			while (!fun()) {
-				std::this_thread::sleep_for(duration_ms / 1000);
+				std::this_thread::sleep_for(duration_ms / 100);
+				const int64_t elapsed_time_us = std::chrono::duration<double, std::micro>(std::chrono::steady_clock::now() -
+								start_time).count();
```
Comment (julianoes) [inline test/mavsdk_tests/autopilot_tester.h:152]: I'm confused about the types here. 

What about?

```suggestion
				std::this_thread::sleep_for(std::chrono::duration_cast<std::chrono::microseconds>(duration_ms)/ 100);
				const auto elapsed_time_ms = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start_time);
```

File: test/mavsdk_tests/autopilot_tester.h [hunk 3]
```diff
@@ -128,24 +128,31 @@ class AutopilotTester
...
 
-				if (iteration++ >= 1000) {
+				if (elapsed_time_us > duration_ms.count() * 1000) {
+					std::cout << "Timeout, waiting for the vehicle for " << elapsed_time_us / 1000000.0 << " seconds" << std::endl;
```
Comment (julianoes) [inline test/mavsdk_tests/autopilot_tester.h:155]: And then:
```suggestion
				if (elapsed_time_ms > duration_ms) {
					std::cout << "Timeout, waiting for the vehicle for " << elapsed_time_ms << " ms\n";
```

**Comment by LorenzMeier** [comment]
I increased the realtime factor for fun to 200 and ran into a manual control test failure. @julianoes the sleep method here will need to be inspected - you need to make sure the C++ automagic is not producing 0 ms sleeps here and you end up with a busy loop. At the end of the day to do this robustly MAVSDK would need to run off the same clock as the sim.

```
	// Climb up for 10 seconds
	for (unsigned i = 0; i < 10 * manual_control_rate_hz; ++i) {
		CHECK(_manual_control->set_manual_control_input(0.f, 0.f, 1.f, 0.f) == ManualControl::Result::Success);
		std::this_thread::sleep_for(adjust_to_lockstep_speed(std::chrono::milliseconds(1000 / manual_control_rate_hz)));
	}
```

**Comment by LorenzMeier** [comment]
Overall for anything relating to time using milliseconds is not a great quantum and mixing ms and us is not helping the readability of the tester code. I would suggest ripping all of that out and normalizing on us as the only unit.

**Comment by LorenzMeier** [comment]
I did a quick experiment changing the quantum to microseconds - that allowed me to run with a much higher realtime factor without test failures, however, it led to starvation on the test setup itself (probably something I changed inconsistently) and so the whole test was slower in the end. 

I'm merging as is as incremental improvement. I'm not convinced I've hit the root cause just yet, but a couple of locations that could cause CI-level failures that are hard to reproduce locally are now more hardened.

**Pull request merged**